chirp/drivers/tk690.py

```diff
@@ -0,0 +1,1211 @@
+        if b is not None:
+            self._del_channel_from_bank(loc, b)
+
+        # adding it
+        self._banks[bank].append(loc - 1)
+
+        # if the update was successful, update in the memmap
+        self._update_bank_memmap()
+
+    def _del_channel_from_bank(self, loc, bank=None):
+        """Remove a channel from a bank, if no bank is specified we find
+        where it is"""
+
+        # if we don't know what bank channel is in,
+        if bank is None:
+            bank = self._get_bank(loc)
+
+        # in case memory entry isn't saved to any bank, no need to delete:
+        if bank is not None:
+            # remove it
+            self._banks[bank].pop(self._banks[bank].index(loc - 1))
+
+        # if the delete was successful update in the memmap
+        self._update_bank_memmap()
+
+    def _update_bank_memmap(self):
+        """This function is called whatever a change is made to a channel
+        or a bank, to update the memmap with the changes that was made"""
+
+        bl = b""
+        bb = b""
+
+        # group_belong index
+        gbi = 0
+        # sort the banks for consistient memory entries
+        self._banks = dict(sorted(self._banks.items()))
+        for bank in self._banks:
+            # check for empty banks
+            if len(self._banks[bank]) == 0:
+                bl += b"\xff\xff"
+                continue
+
+            # channel index inside the bank, starting at 1
+            # aka channel group index
+            cgi = 1
+            for channel in range(0, len(self._banks[bank])):
+                # update bb
+                bb += bytes([cgi, self._banks[bank][channel]])
+                # set the group limits for this group
+                if cgi == 1:
+                    bl += bytes([gbi, len(self._banks[bank])])
+
+                # increment both indexes
+                gbi += 1
+                cgi += 1
+
+        # fill the gaps before write it
+        bb += b"\xff" * 2 * int(self._num_banks - len(bb) / 2)
+        bl += b"\xff" * 2 * int(self._num_banks - len(bl) / 2)
+
+        # update the memmap
+        self._fill(0x1480, bl)
+        self._fill(0x1600, bb)
+
+    def get_settings(self):
+        """Translate the MEM_FORMAT structs into the UI"""
+        button_assignments = RadioSettingGroup("Button Functions",
+                                               "Configurable Button Functions")
+        group = RadioSettings(button_assignments)
+        for buttonName in ASSIGNABLE_BUTTONS:
+            _fullHeadWarning = ""
+            if buttonName in FULL_HEAD_ONLY_BUTTONS:
+                _fullHeadWarning = "  (If Equipped)"
+            rs = RadioSetting(
+              buttonName,
+              "Configured function for "+buttonName+" button"+_fullHeadWarning,
+              RadioSettingValueMap(
+                    BUTTON_FUNCTION_LIST,
+                    self._memobj.button_assignments[buttonName]))
```

Comment:
This should use the non-deprecated `current_index=` pattern instead of the implicit "current value" one it's using here. You should be getting a `FutureWarning` in the log for it.

The newer pattern is safer and allows the UI to handle cases where bad data is stored in the radio so we can highlight the broken setting instead of crashing right here because the value is out of range for the array and thus load no settings.

Reply:
Hmm, can you give an example of using RadioSettingValueMap with a current_index pattern? I couldn't find any examples of it in any other drivers in the repo, and my attempt at it (the following code: 
```python
RadioSettingValueMap(
                    BUTTON_FUNCTION_LIST,
                    current_index=next((i for i, (_, value) in enumerate(BUTTON_FUNCTION_LIST) if value == self._memobj.button_assignments[buttonName]), None)))
```
Produced the following error: `TypeError: RadioSettingValueMap.__init__() got an unexpected keyword argument 'current_index'`
I tried using the same logic but with "user_option" instead of "current_index", and that works, but provides the same FutureWarning as my original logic did.

I tried to open a few of the test images for other radios that used RadioSettingValueMap for their settings to see if any of them handled it differently, (specifically, I've tested the "radtel rt490", the "th7800", and "puxing px888k"), and all three of them produced the same FutureWarning about using current_index instead of value for safety.

Reply:
Sorry, I read this as `RadioSettingValueList` because of the pattern, saw the warning locally when I tried, and didn't realize it was the map. I need to make the same change to map. As you can (probably) see, map uses list as its parent, which is why it's complaining but there's no way to pass the index in.

That's fine, I'll fix that, this, and the others after merge.